pennylane/ops/qubit/matrix_ops.py

```diff
@@ -183,24 +198,45 @@ def __init__(
                 "The control wires must be different from the wires specified to apply the unitary on."
             )
 
-        U = params[0]
-        target_dim = 2 ** len(wires)
-        if len(U) != target_dim:
-            raise ValueError(f"Input unitary must be of shape {(target_dim, target_dim)}")
-
-        # Saving for the circuit drawer
-        self._target_wires = wires
-        self._control_wires = control_wires
-        self.U = U
+        self._hyperparameters = {
+            "u_wires": wires,
+            "control_wires": control_wires,
+            "control_values": control_values,
+        }
```

Comment:
I really like this

Reply:
I like how it provides a place for developers to add hyperparameters, but I also like how there is no restriction on the key names. I think this is important - saying 'you have to name the keys this' and then relying on these names for built-in logic is too much overhead for developers and contributors

Reply:
Everything that _is_ needed for internal logic should be a well-documented property/method instead